### PR TITLE
#600 Vector and point operator micro optimizations

### PR DESCRIPTION
**Pull request by Barsonax**

### CODE REVIEW

**Comment by BobGneu** [comment]
What were the performance improvements to this optimization?

**Comment by SirePi** [comment]
@BobGneu from what I understood in the discussions in the chat, this brings a major speed increase in the execution of the altered operations; I'm just wondering if and how it does affect memory and GC performances due to the high amount of allocations that this change implies, since Vector operations are *a lot* in every update cycle. Maybe I'm just paranoid though 😅

**Comment by Barsonax** [comment]
Well first of all these are structs that are allocated on the stack so I don't think we have anything to do with GC at all here.

Basically the compiler did not optimize the code as far as we would like. We helped the compiler a bit by rewriting the code.

Iam no expert at x86 however but if you look at the assembly it is quite clear that its faster (google the instruction names). See issue #600 for source, IL and assembly code. The old situation was using some pointer like stuff I believe (lea instructions?) and now its all very straight forward.

**Comment by SirePi** [comment]
Ah, perfect then 👍

File: Source/Core/Primitives/Math/Point2.cs [hunk 1]
```diff
@@ -102,104 +102,79 @@ public static float Distance(Point2 left, Point2 right)
 		/// </summary>
 		public static Point2 operator +(Point2 left, Point2 right)
 		{
-			left.X += right.X;
-			left.Y += right.Y;
-			return left;
+			return new Point2(left.X + right.X, left.Y + right.Y);
```
Comment (ilexp) [inline Source/Core/Primitives/Math/Point2.cs]: Can you change the style of the constructor calls in this PR into a multi-line format? Especially with repeating patterns, I think this makes it a lot more readable and easier to spot typos. Consider this:

```csharp
return new Vector4(
	left.X + right.X,
	left.Y + right.Y,
	left.Z + right.Z,
	left.W + right.W);
```

Comment (Barsonax) [reply Source/Core/Primitives/Math/Point2.cs]: Done

**Comment by ilexp** [comment]
Merged!

**Pull request merged**